File: src/policies/slm_heuristic.py
"""SLM-aware heuristic eviction policy (Section 5.1 of the plan).

Hand-tuned policy matching the current SLM-OS Phase 3 implementation.
Uses a priority cascade: workspace before weights, inactive models
before active, then LRU as the final tiebreaker.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .base import EvictionPolicy

if TYPE_CHECKING:
    from src.simulator.block import BlockMeta
    from src.simulator.core import GlobalState, Pool

from src.simulator.block import PoolType


class SLMHeuristicPolicy(EvictionPolicy):
    """Hand-tuned SLM-OS eviction policy.

    Priority cascade:
      1. Evict workspace blocks before weights (cheaper to recreate)
      2. Within workspace, evict LRU
      3. Evict weights from models with no active inferences
      4. Within inactive model weights, evict LRU
      5. Fallback: evict LRU among all remaining candidates
    """

    def __init__(self, active_inferences: dict[int, int] | None = None):
        # Maps model_id -> number of active inference tasks
        self._active_inferences = active_inferences or {}

    def set_active_inferences(self, active: dict[int, int]) -> None:
        """Update the active inference counts (called by simulator)."""
        self._active_inferences = active

    def select_victim(
        self,
        candidates: list[BlockMeta],
        pool: Pool,
        global_state: GlobalState,
    ) -> int:
        # Priority 1: Evict workspace blocks (cheaper to recreate)
        workspace = [
            (i, b) for i, b in enumerate(candidates)
            if b.pool_type == PoolType.WORKSPACE
        ]
        if workspace:
            return self._pick_lru(workspace)

        # Priority 2: Evict weights from models with no active inferences
        inactive = [
            (i, b) for i, b in enumerate(candidates)
            if self._active_inferences.get(b.model_id, 0) == 0
        ]
        if inactive:
            return self._pick_lru(inactive)

        # Priority 3: Evict LRU among remaining
        return self._pick_lru(list(enumerate(candidates)))

    def _pick_lru(self, indexed_blocks: list[tuple[int, BlockMeta]]) -> int:
        """Return the original index of the LRU block."""
        min_time = float("inf")
        victim_idx = indexed_blocks[0][0]
        for idx, block in indexed_blocks:
            if block.last_access_time < min_time:
                min_time = block.last_access_time
                victim_idx = idx
        return victim_idx

    def score(
        self,
        candidates: list[BlockMeta],
        pool: Pool,
        global_state: GlobalState,
    ) -> list[float]:
        """Score candidates by the heuristic priority cascade."""
        scores = [0.0] * len(candidates)
        for i, b in enumerate(candidates):
            score = 0.0
            # Workspace blocks are most evictable
            if b.pool_type == PoolType.WORKSPACE:
                score += 0.6
            # Inactive model blocks are more evictable
            if self._active_inferences.get(b.model_id, 0) == 0:
                score += 0.3
            # Recency tiebreaker (older = more evictable)
            if candidates:
                max_time = max(c.last_access_time for c in candidates)
                min_time = min(c.last_access_time for c in candidates)
                span = max_time - min_time if max_time != min_time else 1
                score += 0.1 * (max_time - b.last_access_time) / span
            scores[i] = score
        return scores

    def name(self) -> str:
        return "SLM-Heuristic"

File: slm_os_integration/src/slm_heuristic.rs
//! Hand-tuned SLM-OS eviction policy.
//!
//! Mirrors `src/policies/slm_heuristic.py` with the same priority cascade:
//!   1. Evict workspace blocks before weights (cheaper to recreate).
//!   2. Within workspace, evict LRU.
//!   3. Evict weights from models with no active inferences.
//!   4. Within inactive model weights, evict LRU.
//!   5. Fallback: evict LRU among all remaining candidates.

use alloc::collections::BTreeMap;
use alloc::vec::Vec;

use crate::block::{BlockMeta, PoolType};
use crate::eviction_policy::EvictionPolicy;

extern crate alloc;

#[derive(Default)]
pub struct SlmHeuristicPolicy {
    /// model_id → number of active inference tasks for that model.
    active_inferences: BTreeMap<u8, u32>,
}

impl SlmHeuristicPolicy {
    pub fn new() -> Self {
        Self { active_inferences: BTreeMap::new() }
    }

    /// Update the active-inference table (called by the runtime when an
    /// inference starts or finishes).
    pub fn set_active_inferences(&mut self, active: BTreeMap<u8, u32>) {
        self.active_inferences = active;
    }

    fn is_active(&self, model_id: u8) -> bool {
        self.active_inferences.get(&model_id).copied().unwrap_or(0) > 0
    }

    /// Return the original index of the LRU block among the supplied
    /// (original_index, block) pairs.
    fn pick_lru(indexed: &[(usize, &BlockMeta)]) -> usize {
        debug_assert!(!indexed.is_empty());
        let mut victim_idx = indexed[0].0;
        let mut min_time = indexed[0].1.last_access_time;
        for (idx, block) in indexed.iter().skip(1) {
            if block.last_access_time < min_time {
                min_time = block.last_access_time;
                victim_idx = *idx;
            }
        }
        victim_idx
    }
}

impl EvictionPolicy for SlmHeuristicPolicy {
    fn select_victim(&mut self, candidates: &[BlockMeta]) -> usize {
        debug_assert!(!candidates.is_empty(), "SLM select_victim on empty list");

        // Priority 1: workspace blocks (cheapest to recreate)
        let workspace: Vec<(usize, &BlockMeta)> = candidates
            .iter()
            .enumerate()
            .filter(|(_, b)| b.pool_type == PoolType::Workspace)
            .collect();
        if !workspace.is_empty() {
            return Self::pick_lru(&workspace);
        }

        // Priority 2: weights from models with no active inferences
        let inactive: Vec<(usize, &BlockMeta)> = candidates
            .iter()
            .enumerate()
            .filter(|(_, b)| !self.is_active(b.model_id))
            .collect();
        if !inactive.is_empty() {
            return Self::pick_lru(&inactive);
        }

        // Priority 3: LRU over all candidates
        let all: Vec<(usize, &BlockMeta)> = candidates.iter().enumerate().collect();
        Self::pick_lru(&all)
    }

    fn score(&mut self, candidates: &[BlockMeta]) -> Vec<f32> {
        if candidates.is_empty() {
            return Vec::new();
        }
        let mut min_time = candidates[0].last_access_time;
        let mut max_time = candidates[0].last_access_time;
        for c in &candidates[1..] {
            if c.last_access_time < min_time { min_time = c.last_access_time; }
            if c.last_access_time > max_time { max_time = c.last_access_time; }
        }
        let span = if max_time == min_time { 1 } else { max_time - min_time };

        candidates
            .iter()
            .map(|b| {
                let mut s = 0.0_f32;
                if b.pool_type == PoolType::Workspace { s += 0.6; }
                if !self.is_active(b.model_id) { s += 0.3; }
                s += 0.1 * (max_time - b.last_access_time) as f32 / span as f32;
                s
            })
            .collect()
    }

    fn name(&self) -> &'static str {
        "SLM-Heuristic"
    }
}
